### PR TITLE
better cache allocation in pytorch engine

### PR DESCRIPTION
**Pull request by grimoire**
runtime intermediate memory usage might lead to OOM if too many cache blocks are using.
Since most intermediate cache are used by `lm_head` in `ModelForCausalLM`, we would leave enough space for the cache and allocate paged cache with rest of the memory.

Tested on V100, prompts with 10k+ input ids.

```python
def main():

    model_path = 'Qwen1.5-7B-Chat'
    backend_config = PytorchEngineConfig(
        session_len=200000,
        cache_max_entry_count=0.99)
    pipe = pipeline(model_path, backend_config=backend_config)

    print('processing...')
    gen_config = GenerationConfig(max_new_tokens=1000)
    result = pipe([prompt], gen_config=gen_config)
    print(result[0].text)
```

### CODE REVIEW

**Comment by grimoire** [comment]
The estimate is based on `vocab_size` and `max_prefill_token_num`, we better expose `max_prefill_token_num` in API.

**Comment by lvhan028** [comment]
`max_prefill_token_num` is already defined in `PyTorchEngineConfig` and `TurbomindEngineConfig`
What does exposing it in API mean?
Pls define small `max_prefill_token_num`

**Comment by grimoire** [comment]
`lmdeploy serve api_server` can not recognize `--max-prefill-token-num`

File: lmdeploy/pytorch/config.py [hunk 1]
```diff
@@ -40,6 +40,7 @@ class CacheConfig:
     num_gpu_blocks: int
     window_size: int = -1
     cache_max_entry_count: float = 0.8
+    max_prefill_token_num: int = 8192
```
Comment (lvhan028) [inline lmdeploy/pytorch/config.py:43]: 为什么PyTorchEngineConfig中的默认值是4096，这里是8192

File: lmdeploy/pytorch/config.py [hunk 2]
```diff
@@ -56,6 +57,7 @@ class ModelConfig:
     sliding_window: int = -1
     dtype: torch.dtype = torch.float16
     multi_query_attention: bool = False
+    vocab_size: int = 40000
```
Comment (lvhan028) [inline lmdeploy/pytorch/config.py:60]: 是不是不该有默认值？

Comment (grimoire) [reply lmdeploy/pytorch/config.py:60]: 是，主要是担心有的模型没有 `vocab_size` 字段，这样可以保底留一部分 runtime memory。40000*4096*7 ~= 1G

**Review by lvhan028 — APPROVED** [review]

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**